internal/client/consensus/grandpa/warp_sync.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2024 ChainSafe Systems (ON)
+// SPDX-License-Identifier: LGPL-3.0-only
+
+package grandpa
+
+import (
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/types"
+	"github.com/ChainSafe/gossamer/internal/primitives/core/hash"
+	"github.com/ChainSafe/gossamer/internal/primitives/runtime"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+const MaxWarpSyncProofSize = 8 * 1024 * 1024
+
+var (
+	errMissingStartBlock      = fmt.Errorf("missing start block")
+	errStartBlockNotFinalized = fmt.Errorf("start block is not finalized")
+)
+
+type BlockState interface {
+	GetHeader(hash common.Hash) (*types.Header, error)
+	BestBlockHeader() (*types.Header, error)
+	GetHeaderByNumber(num uint) (*types.Header, error)
+	GetJustification(hash common.Hash) ([]byte, error)
+}
+
+type GrandpaState interface {
+	GetAuthoritesChangesFromBlock(blockNumber uint) ([]uint, error)
```

Comment:
Should this just be in grandpa service rather than create a separate interface?

Reply:
In Gossamer the grandpa service has as dependency the grandpa state, which is responsible to persist the authorities. Since the goal is to have access to the persisted auth changes then we should use the grandpa state for that.